// @pulumi/aws/appfabric/appAuthorizationConnection.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppAuthorizationConnection = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS AppFabric App Authorization Connection.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.appfabric.AppAuthorizationConnection("example", {
 *     appAuthorizationArn: test.arn,
 *     appBundleArn: arn,
 * });
 * ```
 */
class AppAuthorizationConnection extends pulumi.CustomResource {
    /**
     * Get an existing AppAuthorizationConnection resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new AppAuthorizationConnection(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of AppAuthorizationConnection.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AppAuthorizationConnection.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["app"] = state ? state.app : undefined;
            resourceInputs["appAuthorizationArn"] = state ? state.appAuthorizationArn : undefined;
            resourceInputs["appBundleArn"] = state ? state.appBundleArn : undefined;
            resourceInputs["authRequest"] = state ? state.authRequest : undefined;
            resourceInputs["tenants"] = state ? state.tenants : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.appAuthorizationArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'appAuthorizationArn'");
            }
            if ((!args || args.appBundleArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'appBundleArn'");
            }
            resourceInputs["appAuthorizationArn"] = args ? args.appAuthorizationArn : undefined;
            resourceInputs["appBundleArn"] = args ? args.appBundleArn : undefined;
            resourceInputs["authRequest"] = args ? args.authRequest : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["app"] = undefined /*out*/;
            resourceInputs["tenants"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(AppAuthorizationConnection.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AppAuthorizationConnection = AppAuthorizationConnection;
/** @internal */
AppAuthorizationConnection.__pulumiType = 'aws:appfabric/appAuthorizationConnection:AppAuthorizationConnection';
