// @pulumi/aws/appfabric/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.IngestionDestination = exports.Ingestion = exports.AppBundle = exports.AppAuthorizationConnection = exports.AppAuthorization = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AppAuthorization = null;
utilities.lazyLoad(exports, ["AppAuthorization"], () => require("./appAuthorization"));
exports.AppAuthorizationConnection = null;
utilities.lazyLoad(exports, ["AppAuthorizationConnection"], () => require("./appAuthorizationConnection"));
exports.AppBundle = null;
utilities.lazyLoad(exports, ["AppBundle"], () => require("./appBundle"));
exports.Ingestion = null;
utilities.lazyLoad(exports, ["Ingestion"], () => require("./ingestion"));
exports.IngestionDestination = null;
utilities.lazyLoad(exports, ["IngestionDestination"], () => require("./ingestionDestination"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:appfabric/appAuthorization:AppAuthorization":
                return new exports.AppAuthorization(name, undefined, { urn });
            case "aws:appfabric/appAuthorizationConnection:AppAuthorizationConnection":
                return new exports.AppAuthorizationConnection(name, undefined, { urn });
            case "aws:appfabric/appBundle:AppBundle":
                return new exports.AppBundle(name, undefined, { urn });
            case "aws:appfabric/ingestion:Ingestion":
                return new exports.Ingestion(name, undefined, { urn });
            case "aws:appfabric/ingestionDestination:IngestionDestination":
                return new exports.IngestionDestination(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "appfabric/appAuthorization", _module);
pulumi.runtime.registerResourceModule("aws", "appfabric/appAuthorizationConnection", _module);
pulumi.runtime.registerResourceModule("aws", "appfabric/appBundle", _module);
pulumi.runtime.registerResourceModule("aws", "appfabric/ingestion", _module);
pulumi.runtime.registerResourceModule("aws", "appfabric/ingestionDestination", _module);
