// @pulumi/aws/appfabric/ingestionDestination.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.IngestionDestination = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS AppFabric Ingestion Destination.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.appfabric.IngestionDestination("example", {
 *     appBundleArn: exampleAwsAppfabricAppBundle.arn,
 *     ingestionArn: exampleAwsAppfabricIngestion.arn,
 *     processingConfiguration: {
 *         auditLog: {
 *             format: "json",
 *             schema: "raw",
 *         },
 *     },
 *     destinationConfiguration: {
 *         auditLog: {
 *             destination: {
 *                 s3Bucket: {
 *                     bucketName: exampleAwsS3Bucket.bucket,
 *                 },
 *             },
 *         },
 *     },
 * });
 * ```
 */
class IngestionDestination extends pulumi.CustomResource {
    /**
     * Get an existing IngestionDestination resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new IngestionDestination(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of IngestionDestination.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === IngestionDestination.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["appBundleArn"] = state ? state.appBundleArn : undefined;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["destinationConfiguration"] = state ? state.destinationConfiguration : undefined;
            resourceInputs["ingestionArn"] = state ? state.ingestionArn : undefined;
            resourceInputs["processingConfiguration"] = state ? state.processingConfiguration : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.appBundleArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'appBundleArn'");
            }
            if ((!args || args.ingestionArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'ingestionArn'");
            }
            resourceInputs["appBundleArn"] = args ? args.appBundleArn : undefined;
            resourceInputs["destinationConfiguration"] = args ? args.destinationConfiguration : undefined;
            resourceInputs["ingestionArn"] = args ? args.ingestionArn : undefined;
            resourceInputs["processingConfiguration"] = args ? args.processingConfiguration : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(IngestionDestination.__pulumiType, name, resourceInputs, opts);
    }
}
exports.IngestionDestination = IngestionDestination;
/** @internal */
IngestionDestination.__pulumiType = 'aws:appfabric/ingestionDestination:IngestionDestination';
