// @pulumi/aws/medialive/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiplexProgram = exports.Multiplex = exports.InputSecurityGroup = exports.Input = exports.getInputOutput = exports.getInput = exports.Channel = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Channel = null;
utilities.lazyLoad(exports, ["Channel"], () => require("./channel"));
exports.getInput = null;
exports.getInputOutput = null;
utilities.lazyLoad(exports, ["getInput", "getInputOutput"], () => require("./getInput"));
exports.Input = null;
utilities.lazyLoad(exports, ["Input"], () => require("./input"));
exports.InputSecurityGroup = null;
utilities.lazyLoad(exports, ["InputSecurityGroup"], () => require("./inputSecurityGroup"));
exports.Multiplex = null;
utilities.lazyLoad(exports, ["Multiplex"], () => require("./multiplex"));
exports.MultiplexProgram = null;
utilities.lazyLoad(exports, ["MultiplexProgram"], () => require("./multiplexProgram"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:medialive/channel:Channel":
                return new exports.Channel(name, undefined, { urn });
            case "aws:medialive/input:Input":
                return new exports.Input(name, undefined, { urn });
            case "aws:medialive/inputSecurityGroup:InputSecurityGroup":
                return new exports.InputSecurityGroup(name, undefined, { urn });
            case "aws:medialive/multiplex:Multiplex":
                return new exports.Multiplex(name, undefined, { urn });
            case "aws:medialive/multiplexProgram:MultiplexProgram":
                return new exports.MultiplexProgram(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "medialive/channel", _module);
pulumi.runtime.registerResourceModule("aws", "medialive/input", _module);
pulumi.runtime.registerResourceModule("aws", "medialive/inputSecurityGroup", _module);
pulumi.runtime.registerResourceModule("aws", "medialive/multiplex", _module);
pulumi.runtime.registerResourceModule("aws", "medialive/multiplexProgram", _module);
