// @pulumi/aws/controltower/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.LandingZone = exports.getControlsOutput = exports.getControls = exports.ControlTowerControl = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.ControlTowerControl = null;
utilities.lazyLoad(exports, ["ControlTowerControl"], () => require("./controlTowerControl"));
exports.getControls = null;
exports.getControlsOutput = null;
utilities.lazyLoad(exports, ["getControls", "getControlsOutput"], () => require("./getControls"));
exports.LandingZone = null;
utilities.lazyLoad(exports, ["LandingZone"], () => require("./landingZone"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:controltower/controlTowerControl:ControlTowerControl":
                return new exports.ControlTowerControl(name, undefined, { urn });
            case "aws:controltower/landingZone:LandingZone":
                return new exports.LandingZone(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "controltower/controlTowerControl", _module);
pulumi.runtime.registerResourceModule("aws", "controltower/landingZone", _module);
