// @pulumi/aws/finspace/kxCluster.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.KxCluster = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS FinSpace Kx Cluster.
 *
 * ## Example Usage
 *
 * ## Import
 *
 * Using `pulumi import`, import an AWS FinSpace Kx Cluster using the `id` (environment ID and cluster name, comma-delimited). For example:
 *
 * ```sh
 * $ pulumi import aws:finspace/kxCluster:KxCluster example n3ceo7wqxoxcti5tujqwzs,my-tf-kx-cluster
 * ```
 */
class KxCluster extends pulumi.CustomResource {
    /**
     * Get an existing KxCluster resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new KxCluster(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of KxCluster.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === KxCluster.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["autoScalingConfiguration"] = state ? state.autoScalingConfiguration : undefined;
            resourceInputs["availabilityZoneId"] = state ? state.availabilityZoneId : undefined;
            resourceInputs["azMode"] = state ? state.azMode : undefined;
            resourceInputs["cacheStorageConfigurations"] = state ? state.cacheStorageConfigurations : undefined;
            resourceInputs["capacityConfiguration"] = state ? state.capacityConfiguration : undefined;
            resourceInputs["code"] = state ? state.code : undefined;
            resourceInputs["commandLineArguments"] = state ? state.commandLineArguments : undefined;
            resourceInputs["createdTimestamp"] = state ? state.createdTimestamp : undefined;
            resourceInputs["databases"] = state ? state.databases : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["environmentId"] = state ? state.environmentId : undefined;
            resourceInputs["executionRole"] = state ? state.executionRole : undefined;
            resourceInputs["initializationScript"] = state ? state.initializationScript : undefined;
            resourceInputs["lastModifiedTimestamp"] = state ? state.lastModifiedTimestamp : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["releaseLabel"] = state ? state.releaseLabel : undefined;
            resourceInputs["savedownStorageConfiguration"] = state ? state.savedownStorageConfiguration : undefined;
            resourceInputs["scalingGroupConfiguration"] = state ? state.scalingGroupConfiguration : undefined;
            resourceInputs["status"] = state ? state.status : undefined;
            resourceInputs["statusReason"] = state ? state.statusReason : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
            resourceInputs["tickerplantLogConfigurations"] = state ? state.tickerplantLogConfigurations : undefined;
            resourceInputs["type"] = state ? state.type : undefined;
            resourceInputs["vpcConfiguration"] = state ? state.vpcConfiguration : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.azMode === undefined) && !opts.urn) {
                throw new Error("Missing required property 'azMode'");
            }
            if ((!args || args.environmentId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'environmentId'");
            }
            if ((!args || args.releaseLabel === undefined) && !opts.urn) {
                throw new Error("Missing required property 'releaseLabel'");
            }
            if ((!args || args.type === undefined) && !opts.urn) {
                throw new Error("Missing required property 'type'");
            }
            if ((!args || args.vpcConfiguration === undefined) && !opts.urn) {
                throw new Error("Missing required property 'vpcConfiguration'");
            }
            resourceInputs["autoScalingConfiguration"] = args ? args.autoScalingConfiguration : undefined;
            resourceInputs["availabilityZoneId"] = args ? args.availabilityZoneId : undefined;
            resourceInputs["azMode"] = args ? args.azMode : undefined;
            resourceInputs["cacheStorageConfigurations"] = args ? args.cacheStorageConfigurations : undefined;
            resourceInputs["capacityConfiguration"] = args ? args.capacityConfiguration : undefined;
            resourceInputs["code"] = args ? args.code : undefined;
            resourceInputs["commandLineArguments"] = args ? args.commandLineArguments : undefined;
            resourceInputs["databases"] = args ? args.databases : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["environmentId"] = args ? args.environmentId : undefined;
            resourceInputs["executionRole"] = args ? args.executionRole : undefined;
            resourceInputs["initializationScript"] = args ? args.initializationScript : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["releaseLabel"] = args ? args.releaseLabel : undefined;
            resourceInputs["savedownStorageConfiguration"] = args ? args.savedownStorageConfiguration : undefined;
            resourceInputs["scalingGroupConfiguration"] = args ? args.scalingGroupConfiguration : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["tickerplantLogConfigurations"] = args ? args.tickerplantLogConfigurations : undefined;
            resourceInputs["type"] = args ? args.type : undefined;
            resourceInputs["vpcConfiguration"] = args ? args.vpcConfiguration : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["createdTimestamp"] = undefined /*out*/;
            resourceInputs["lastModifiedTimestamp"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["statusReason"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(KxCluster.__pulumiType, name, resourceInputs, opts);
    }
}
exports.KxCluster = KxCluster;
/** @internal */
KxCluster.__pulumiType = 'aws:finspace/kxCluster:KxCluster';
