// @pulumi/synced-folder/azureBlobFolder.js
"use strict";
// *** WARNING: this file was generated by Pulumi SDK Generator. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureBlobFolder = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
class AzureBlobFolder extends pulumi.ComponentResource {
    /**
     * Returns true if the given object is an instance of AzureBlobFolder.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AzureBlobFolder.__pulumiType;
    }
    /**
     * Create a AzureBlobFolder resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.containerName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'containerName'");
            }
            if ((!args || args.path === undefined) && !opts.urn) {
                throw new Error("Missing required property 'path'");
            }
            if ((!args || args.resourceGroupName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'resourceGroupName'");
            }
            if ((!args || args.storageAccountName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'storageAccountName'");
            }
            resourceInputs["containerName"] = args ? args.containerName : undefined;
            resourceInputs["managedObjects"] = args ? args.managedObjects : undefined;
            resourceInputs["path"] = args ? args.path : undefined;
            resourceInputs["resourceGroupName"] = args ? args.resourceGroupName : undefined;
            resourceInputs["storageAccountName"] = args ? args.storageAccountName : undefined;
        }
        else {
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(AzureBlobFolder.__pulumiType, name, resourceInputs, opts, true /*remote*/);
    }
}
exports.AzureBlobFolder = AzureBlobFolder;
/** @internal */
AzureBlobFolder.__pulumiType = 'synced-folder:index:AzureBlobFolder';
