// @pulumi/synced-folder/s3bucketFolder.js
"use strict";
// *** WARNING: this file was generated by Pulumi SDK Generator. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketFolder = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
class S3BucketFolder extends pulumi.ComponentResource {
    /**
     * Returns true if the given object is an instance of S3BucketFolder.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === S3BucketFolder.__pulumiType;
    }
    /**
     * Create a S3BucketFolder resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.acl === undefined) && !opts.urn) {
                throw new Error("Missing required property 'acl'");
            }
            if ((!args || args.bucketName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'bucketName'");
            }
            if ((!args || args.path === undefined) && !opts.urn) {
                throw new Error("Missing required property 'path'");
            }
            resourceInputs["acl"] = args ? args.acl : undefined;
            resourceInputs["bucketName"] = args ? args.bucketName : undefined;
            resourceInputs["managedObjects"] = args ? args.managedObjects : undefined;
            resourceInputs["path"] = args ? args.path : undefined;
        }
        else {
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(S3BucketFolder.__pulumiType, name, resourceInputs, opts, true /*remote*/);
    }
}
exports.S3BucketFolder = S3BucketFolder;
/** @internal */
S3BucketFolder.__pulumiType = 'synced-folder:index:S3BucketFolder';
