// @pulumi/aws/timestreamwrite/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = exports.getTableOutput = exports.getTable = exports.getDatabaseOutput = exports.getDatabase = exports.Database = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Database = null;
utilities.lazyLoad(exports, ["Database"], () => require("./database"));
exports.getDatabase = null;
exports.getDatabaseOutput = null;
utilities.lazyLoad(exports, ["getDatabase", "getDatabaseOutput"], () => require("./getDatabase"));
exports.getTable = null;
exports.getTableOutput = null;
utilities.lazyLoad(exports, ["getTable", "getTableOutput"], () => require("./getTable"));
exports.Table = null;
utilities.lazyLoad(exports, ["Table"], () => require("./table"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:timestreamwrite/database:Database":
                return new exports.Database(name, undefined, { urn });
            case "aws:timestreamwrite/table:Table":
                return new exports.Table(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "timestreamwrite/database", _module);
pulumi.runtime.registerResourceModule("aws", "timestreamwrite/table", _module);
