// @pulumi/aws/dataexchange/eventAction.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventAction = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS Data Exchange Event Action.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.dataexchange.EventAction("example", {
 *     event: {
 *         revisionPublished: {
 *             dataSetId: exampleAwsDataexchangeDataSet.id,
 *         },
 *     },
 *     action: {
 *         exportRevisionToS3: {
 *             revisionDestination: {
 *                 bucket: exampleAwsS3Bucket.bucket,
 *                 keyPattern: "${Revision.CreatedAt}/${Asset.Name}",
 *             },
 *             encryption: {
 *                 type: "aws:kms",
 *                 kmsKeyArn: exampleAwsKmsKey.arn,
 *             },
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import Data Exchange Event Action using the id. For example:
 *
 * ```sh
 * $ pulumi import aws:dataexchange/eventAction:EventAction example example-event-action-id
 * ```
 */
class EventAction extends pulumi.CustomResource {
    /**
     * Get an existing EventAction resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new EventAction(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EventAction.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EventAction.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["action"] = state ? state.action : undefined;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["createdAt"] = state ? state.createdAt : undefined;
            resourceInputs["event"] = state ? state.event : undefined;
            resourceInputs["updatedAt"] = state ? state.updatedAt : undefined;
        }
        else {
            const args = argsOrState;
            resourceInputs["action"] = args ? args.action : undefined;
            resourceInputs["event"] = args ? args.event : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["updatedAt"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(EventAction.__pulumiType, name, resourceInputs, opts);
    }
}
exports.EventAction = EventAction;
/** @internal */
EventAction.__pulumiType = 'aws:dataexchange/eventAction:EventAction';
