// @pulumi/aws/lex/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2modelsSlotType = exports.V2modelsSlot = exports.V2modelsIntent = exports.V2modelsBotVersion = exports.V2modelsBotLocale = exports.V2modelsBot = exports.SlotType = exports.Intent = exports.getSlotTypeOutput = exports.getSlotType = exports.getIntentOutput = exports.getIntent = exports.getBotAliasOutput = exports.getBotAlias = exports.getBotOutput = exports.getBot = exports.BotAlias = exports.Bot = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Bot = null;
utilities.lazyLoad(exports, ["Bot"], () => require("./bot"));
exports.BotAlias = null;
utilities.lazyLoad(exports, ["BotAlias"], () => require("./botAlias"));
exports.getBot = null;
exports.getBotOutput = null;
utilities.lazyLoad(exports, ["getBot", "getBotOutput"], () => require("./getBot"));
exports.getBotAlias = null;
exports.getBotAliasOutput = null;
utilities.lazyLoad(exports, ["getBotAlias", "getBotAliasOutput"], () => require("./getBotAlias"));
exports.getIntent = null;
exports.getIntentOutput = null;
utilities.lazyLoad(exports, ["getIntent", "getIntentOutput"], () => require("./getIntent"));
exports.getSlotType = null;
exports.getSlotTypeOutput = null;
utilities.lazyLoad(exports, ["getSlotType", "getSlotTypeOutput"], () => require("./getSlotType"));
exports.Intent = null;
utilities.lazyLoad(exports, ["Intent"], () => require("./intent"));
exports.SlotType = null;
utilities.lazyLoad(exports, ["SlotType"], () => require("./slotType"));
exports.V2modelsBot = null;
utilities.lazyLoad(exports, ["V2modelsBot"], () => require("./v2modelsBot"));
exports.V2modelsBotLocale = null;
utilities.lazyLoad(exports, ["V2modelsBotLocale"], () => require("./v2modelsBotLocale"));
exports.V2modelsBotVersion = null;
utilities.lazyLoad(exports, ["V2modelsBotVersion"], () => require("./v2modelsBotVersion"));
exports.V2modelsIntent = null;
utilities.lazyLoad(exports, ["V2modelsIntent"], () => require("./v2modelsIntent"));
exports.V2modelsSlot = null;
utilities.lazyLoad(exports, ["V2modelsSlot"], () => require("./v2modelsSlot"));
exports.V2modelsSlotType = null;
utilities.lazyLoad(exports, ["V2modelsSlotType"], () => require("./v2modelsSlotType"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:lex/bot:Bot":
                return new exports.Bot(name, undefined, { urn });
            case "aws:lex/botAlias:BotAlias":
                return new exports.BotAlias(name, undefined, { urn });
            case "aws:lex/intent:Intent":
                return new exports.Intent(name, undefined, { urn });
            case "aws:lex/slotType:SlotType":
                return new exports.SlotType(name, undefined, { urn });
            case "aws:lex/v2modelsBot:V2modelsBot":
                return new exports.V2modelsBot(name, undefined, { urn });
            case "aws:lex/v2modelsBotLocale:V2modelsBotLocale":
                return new exports.V2modelsBotLocale(name, undefined, { urn });
            case "aws:lex/v2modelsBotVersion:V2modelsBotVersion":
                return new exports.V2modelsBotVersion(name, undefined, { urn });
            case "aws:lex/v2modelsIntent:V2modelsIntent":
                return new exports.V2modelsIntent(name, undefined, { urn });
            case "aws:lex/v2modelsSlot:V2modelsSlot":
                return new exports.V2modelsSlot(name, undefined, { urn });
            case "aws:lex/v2modelsSlotType:V2modelsSlotType":
                return new exports.V2modelsSlotType(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "lex/bot", _module);
pulumi.runtime.registerResourceModule("aws", "lex/botAlias", _module);
pulumi.runtime.registerResourceModule("aws", "lex/intent", _module);
pulumi.runtime.registerResourceModule("aws", "lex/slotType", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsBot", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsBotLocale", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsBotVersion", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsIntent", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsSlot", _module);
pulumi.runtime.registerResourceModule("aws", "lex/v2modelsSlotType", _module);
