// @pulumi/aws/lex/v2modelsBotVersion.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2modelsBotVersion = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS Lex V2 Models Bot Version.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.lex.V2modelsBotVersion("test", {
 *     botId: testAwsLexv2modelsBot.id,
 *     localeSpecification: {
 *         en_US: {
 *             sourceBotVersion: "DRAFT",
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import Lex V2 Models Bot Version using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:lex/v2modelsBotVersion:V2modelsBotVersion example id-12345678,1
 * ```
 */
class V2modelsBotVersion extends pulumi.CustomResource {
    /**
     * Get an existing V2modelsBotVersion resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new V2modelsBotVersion(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of V2modelsBotVersion.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === V2modelsBotVersion.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["botId"] = state ? state.botId : undefined;
            resourceInputs["botVersion"] = state ? state.botVersion : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["localeSpecification"] = state ? state.localeSpecification : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.botId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'botId'");
            }
            if ((!args || args.localeSpecification === undefined) && !opts.urn) {
                throw new Error("Missing required property 'localeSpecification'");
            }
            resourceInputs["botId"] = args ? args.botId : undefined;
            resourceInputs["botVersion"] = args ? args.botVersion : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["localeSpecification"] = args ? args.localeSpecification : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(V2modelsBotVersion.__pulumiType, name, resourceInputs, opts);
    }
}
exports.V2modelsBotVersion = V2modelsBotVersion;
/** @internal */
V2modelsBotVersion.__pulumiType = 'aws:lex/v2modelsBotVersion:V2modelsBotVersion';
