// @pulumi/aws/lex/v2modelsSlot.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2modelsSlot = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * ## Import
 *
 * Using `pulumi import`, import Lex V2 Models Slot using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:lex/v2modelsSlot:V2modelsSlot example bot-1234,1,intent-5678,en-US,slot-9012
 * ```
 */
class V2modelsSlot extends pulumi.CustomResource {
    /**
     * Get an existing V2modelsSlot resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new V2modelsSlot(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of V2modelsSlot.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === V2modelsSlot.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["botId"] = state ? state.botId : undefined;
            resourceInputs["botVersion"] = state ? state.botVersion : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["intentId"] = state ? state.intentId : undefined;
            resourceInputs["localeId"] = state ? state.localeId : undefined;
            resourceInputs["multipleValuesSettings"] = state ? state.multipleValuesSettings : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["obfuscationSettings"] = state ? state.obfuscationSettings : undefined;
            resourceInputs["slotId"] = state ? state.slotId : undefined;
            resourceInputs["slotTypeId"] = state ? state.slotTypeId : undefined;
            resourceInputs["subSlotSettings"] = state ? state.subSlotSettings : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
            resourceInputs["valueElicitationSetting"] = state ? state.valueElicitationSetting : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.botId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'botId'");
            }
            if ((!args || args.botVersion === undefined) && !opts.urn) {
                throw new Error("Missing required property 'botVersion'");
            }
            if ((!args || args.intentId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'intentId'");
            }
            if ((!args || args.localeId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'localeId'");
            }
            resourceInputs["botId"] = args ? args.botId : undefined;
            resourceInputs["botVersion"] = args ? args.botVersion : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["intentId"] = args ? args.intentId : undefined;
            resourceInputs["localeId"] = args ? args.localeId : undefined;
            resourceInputs["multipleValuesSettings"] = args ? args.multipleValuesSettings : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["obfuscationSettings"] = args ? args.obfuscationSettings : undefined;
            resourceInputs["slotTypeId"] = args ? args.slotTypeId : undefined;
            resourceInputs["subSlotSettings"] = args ? args.subSlotSettings : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["valueElicitationSetting"] = args ? args.valueElicitationSetting : undefined;
            resourceInputs["slotId"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(V2modelsSlot.__pulumiType, name, resourceInputs, opts);
    }
}
exports.V2modelsSlot = V2modelsSlot;
/** @internal */
V2modelsSlot.__pulumiType = 'aws:lex/v2modelsSlot:V2modelsSlot';
