// @pulumi/aws/vpc/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityGroupVpcAssociation = exports.SecurityGroupIngressRule = exports.SecurityGroupEgressRule = exports.getSecurityGroupRulesOutput = exports.getSecurityGroupRules = exports.getSecurityGroupRuleOutput = exports.getSecurityGroupRule = exports.EndpointServicePrivateDnsVerification = exports.EndpointPrivateDns = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.EndpointPrivateDns = null;
utilities.lazyLoad(exports, ["EndpointPrivateDns"], () => require("./endpointPrivateDns"));
exports.EndpointServicePrivateDnsVerification = null;
utilities.lazyLoad(exports, ["EndpointServicePrivateDnsVerification"], () => require("./endpointServicePrivateDnsVerification"));
exports.getSecurityGroupRule = null;
exports.getSecurityGroupRuleOutput = null;
utilities.lazyLoad(exports, ["getSecurityGroupRule", "getSecurityGroupRuleOutput"], () => require("./getSecurityGroupRule"));
exports.getSecurityGroupRules = null;
exports.getSecurityGroupRulesOutput = null;
utilities.lazyLoad(exports, ["getSecurityGroupRules", "getSecurityGroupRulesOutput"], () => require("./getSecurityGroupRules"));
exports.SecurityGroupEgressRule = null;
utilities.lazyLoad(exports, ["SecurityGroupEgressRule"], () => require("./securityGroupEgressRule"));
exports.SecurityGroupIngressRule = null;
utilities.lazyLoad(exports, ["SecurityGroupIngressRule"], () => require("./securityGroupIngressRule"));
exports.SecurityGroupVpcAssociation = null;
utilities.lazyLoad(exports, ["SecurityGroupVpcAssociation"], () => require("./securityGroupVpcAssociation"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:vpc/endpointPrivateDns:EndpointPrivateDns":
                return new exports.EndpointPrivateDns(name, undefined, { urn });
            case "aws:vpc/endpointServicePrivateDnsVerification:EndpointServicePrivateDnsVerification":
                return new exports.EndpointServicePrivateDnsVerification(name, undefined, { urn });
            case "aws:vpc/securityGroupEgressRule:SecurityGroupEgressRule":
                return new exports.SecurityGroupEgressRule(name, undefined, { urn });
            case "aws:vpc/securityGroupIngressRule:SecurityGroupIngressRule":
                return new exports.SecurityGroupIngressRule(name, undefined, { urn });
            case "aws:vpc/securityGroupVpcAssociation:SecurityGroupVpcAssociation":
                return new exports.SecurityGroupVpcAssociation(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "vpc/endpointPrivateDns", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/endpointServicePrivateDnsVerification", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupEgressRule", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupIngressRule", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupVpcAssociation", _module);
