// @pulumi/aws/globalaccelerator/customRoutingEndpointGroup.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomRoutingEndpointGroup = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Global Accelerator custom routing endpoint group.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.globalaccelerator.CustomRoutingEndpointGroup("example", {
 *     listenerArn: exampleAwsGlobalacceleratorCustomRoutingListener.arn,
 *     destinationConfigurations: [{
 *         fromPort: 80,
 *         toPort: 8080,
 *         protocols: ["TCP"],
 *     }],
 *     endpointConfigurations: [{
 *         endpointId: exampleAwsSubnet.id,
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import Global Accelerator custom routing endpoint groups using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:globalaccelerator/customRoutingEndpointGroup:CustomRoutingEndpointGroup example arn:aws:globalaccelerator::111111111111:accelerator/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/listener/xxxxxxx/endpoint-group/xxxxxxxx
 * ```
 */
class CustomRoutingEndpointGroup extends pulumi.CustomResource {
    /**
     * Get an existing CustomRoutingEndpointGroup resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CustomRoutingEndpointGroup(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of CustomRoutingEndpointGroup.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CustomRoutingEndpointGroup.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["destinationConfigurations"] = state ? state.destinationConfigurations : undefined;
            resourceInputs["endpointConfigurations"] = state ? state.endpointConfigurations : undefined;
            resourceInputs["endpointGroupRegion"] = state ? state.endpointGroupRegion : undefined;
            resourceInputs["listenerArn"] = state ? state.listenerArn : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.destinationConfigurations === undefined) && !opts.urn) {
                throw new Error("Missing required property 'destinationConfigurations'");
            }
            if ((!args || args.listenerArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'listenerArn'");
            }
            resourceInputs["destinationConfigurations"] = args ? args.destinationConfigurations : undefined;
            resourceInputs["endpointConfigurations"] = args ? args.endpointConfigurations : undefined;
            resourceInputs["endpointGroupRegion"] = args ? args.endpointGroupRegion : undefined;
            resourceInputs["listenerArn"] = args ? args.listenerArn : undefined;
            resourceInputs["arn"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(CustomRoutingEndpointGroup.__pulumiType, name, resourceInputs, opts);
    }
}
exports.CustomRoutingEndpointGroup = CustomRoutingEndpointGroup;
/** @internal */
CustomRoutingEndpointGroup.__pulumiType = 'aws:globalaccelerator/customRoutingEndpointGroup:CustomRoutingEndpointGroup';
