// @pulumi/aws/guardduty/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThreatIntelSet = exports.PublishingDestination = exports.OrganizationConfigurationFeature = exports.OrganizationConfiguration = exports.OrganizationAdminAccount = exports.MemberDetectorFeature = exports.Member = exports.MalwareProtectionPlan = exports.IPSet = exports.InviteAccepter = exports.getFindingIdsOutput = exports.getFindingIds = exports.getDetectorOutput = exports.getDetector = exports.Filter = exports.DetectorFeature = exports.Detector = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Detector = null;
utilities.lazyLoad(exports, ["Detector"], () => require("./detector"));
exports.DetectorFeature = null;
utilities.lazyLoad(exports, ["DetectorFeature"], () => require("./detectorFeature"));
exports.Filter = null;
utilities.lazyLoad(exports, ["Filter"], () => require("./filter"));
exports.getDetector = null;
exports.getDetectorOutput = null;
utilities.lazyLoad(exports, ["getDetector", "getDetectorOutput"], () => require("./getDetector"));
exports.getFindingIds = null;
exports.getFindingIdsOutput = null;
utilities.lazyLoad(exports, ["getFindingIds", "getFindingIdsOutput"], () => require("./getFindingIds"));
exports.InviteAccepter = null;
utilities.lazyLoad(exports, ["InviteAccepter"], () => require("./inviteAccepter"));
exports.IPSet = null;
utilities.lazyLoad(exports, ["IPSet"], () => require("./ipset"));
exports.MalwareProtectionPlan = null;
utilities.lazyLoad(exports, ["MalwareProtectionPlan"], () => require("./malwareProtectionPlan"));
exports.Member = null;
utilities.lazyLoad(exports, ["Member"], () => require("./member"));
exports.MemberDetectorFeature = null;
utilities.lazyLoad(exports, ["MemberDetectorFeature"], () => require("./memberDetectorFeature"));
exports.OrganizationAdminAccount = null;
utilities.lazyLoad(exports, ["OrganizationAdminAccount"], () => require("./organizationAdminAccount"));
exports.OrganizationConfiguration = null;
utilities.lazyLoad(exports, ["OrganizationConfiguration"], () => require("./organizationConfiguration"));
exports.OrganizationConfigurationFeature = null;
utilities.lazyLoad(exports, ["OrganizationConfigurationFeature"], () => require("./organizationConfigurationFeature"));
exports.PublishingDestination = null;
utilities.lazyLoad(exports, ["PublishingDestination"], () => require("./publishingDestination"));
exports.ThreatIntelSet = null;
utilities.lazyLoad(exports, ["ThreatIntelSet"], () => require("./threatIntelSet"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:guardduty/detector:Detector":
                return new exports.Detector(name, undefined, { urn });
            case "aws:guardduty/detectorFeature:DetectorFeature":
                return new exports.DetectorFeature(name, undefined, { urn });
            case "aws:guardduty/filter:Filter":
                return new exports.Filter(name, undefined, { urn });
            case "aws:guardduty/iPSet:IPSet":
                return new exports.IPSet(name, undefined, { urn });
            case "aws:guardduty/inviteAccepter:InviteAccepter":
                return new exports.InviteAccepter(name, undefined, { urn });
            case "aws:guardduty/malwareProtectionPlan:MalwareProtectionPlan":
                return new exports.MalwareProtectionPlan(name, undefined, { urn });
            case "aws:guardduty/member:Member":
                return new exports.Member(name, undefined, { urn });
            case "aws:guardduty/memberDetectorFeature:MemberDetectorFeature":
                return new exports.MemberDetectorFeature(name, undefined, { urn });
            case "aws:guardduty/organizationAdminAccount:OrganizationAdminAccount":
                return new exports.OrganizationAdminAccount(name, undefined, { urn });
            case "aws:guardduty/organizationConfiguration:OrganizationConfiguration":
                return new exports.OrganizationConfiguration(name, undefined, { urn });
            case "aws:guardduty/organizationConfigurationFeature:OrganizationConfigurationFeature":
                return new exports.OrganizationConfigurationFeature(name, undefined, { urn });
            case "aws:guardduty/publishingDestination:PublishingDestination":
                return new exports.PublishingDestination(name, undefined, { urn });
            case "aws:guardduty/threatIntelSet:ThreatIntelSet":
                return new exports.ThreatIntelSet(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "guardduty/detector", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/detectorFeature", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/filter", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/iPSet", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/inviteAccepter", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/malwareProtectionPlan", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/member", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/memberDetectorFeature", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/organizationAdminAccount", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/organizationConfiguration", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/organizationConfigurationFeature", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/publishingDestination", _module);
pulumi.runtime.registerResourceModule("aws", "guardduty/threatIntelSet", _module);
