// @pulumi/aws/sagemaker/hub.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hub = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a SageMaker AI Hub resource.
 *
 * ## Example Usage
 *
 * ### Basic usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.sagemaker.Hub("example", {
 *     hubName: "example",
 *     hubDescription: "example",
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import SageMaker AI Hubs using the `name`. For example:
 *
 * ```sh
 * $ pulumi import aws:sagemaker/hub:Hub test_hub my-code-repo
 * ```
 */
class Hub extends pulumi.CustomResource {
    /**
     * Get an existing Hub resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Hub(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Hub.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Hub.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["hubDescription"] = state ? state.hubDescription : undefined;
            resourceInputs["hubDisplayName"] = state ? state.hubDisplayName : undefined;
            resourceInputs["hubName"] = state ? state.hubName : undefined;
            resourceInputs["hubSearchKeywords"] = state ? state.hubSearchKeywords : undefined;
            resourceInputs["s3StorageConfig"] = state ? state.s3StorageConfig : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.hubDescription === undefined) && !opts.urn) {
                throw new Error("Missing required property 'hubDescription'");
            }
            if ((!args || args.hubName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'hubName'");
            }
            resourceInputs["hubDescription"] = args ? args.hubDescription : undefined;
            resourceInputs["hubDisplayName"] = args ? args.hubDisplayName : undefined;
            resourceInputs["hubName"] = args ? args.hubName : undefined;
            resourceInputs["hubSearchKeywords"] = args ? args.hubSearchKeywords : undefined;
            resourceInputs["s3StorageConfig"] = args ? args.s3StorageConfig : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Hub.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Hub = Hub;
/** @internal */
Hub.__pulumiType = 'aws:sagemaker/hub:Hub';
