// @pulumi/aws/backup/logicallyAirGappedVault.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogicallyAirGappedVault = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS Backup Logically Air Gapped Vault.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.backup.LogicallyAirGappedVault("example", {
 *     name: "lag-example-vault",
 *     maxRetentionDays: 7,
 *     minRetentionDays: 7,
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import Backup Logically Air Gapped Vault using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:backup/logicallyAirGappedVault:LogicallyAirGappedVault example lag-example-vault
 * ```
 */
class LogicallyAirGappedVault extends pulumi.CustomResource {
    /**
     * Get an existing LogicallyAirGappedVault resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new LogicallyAirGappedVault(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of LogicallyAirGappedVault.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === LogicallyAirGappedVault.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["maxRetentionDays"] = state ? state.maxRetentionDays : undefined;
            resourceInputs["minRetentionDays"] = state ? state.minRetentionDays : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.maxRetentionDays === undefined) && !opts.urn) {
                throw new Error("Missing required property 'maxRetentionDays'");
            }
            if ((!args || args.minRetentionDays === undefined) && !opts.urn) {
                throw new Error("Missing required property 'minRetentionDays'");
            }
            resourceInputs["maxRetentionDays"] = args ? args.maxRetentionDays : undefined;
            resourceInputs["minRetentionDays"] = args ? args.minRetentionDays : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(LogicallyAirGappedVault.__pulumiType, name, resourceInputs, opts);
    }
}
exports.LogicallyAirGappedVault = LogicallyAirGappedVault;
/** @internal */
LogicallyAirGappedVault.__pulumiType = 'aws:backup/logicallyAirGappedVault:LogicallyAirGappedVault';
