// @pulumi/aws/codecatalyst/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceRepository = exports.Project = exports.getDevEnvironmentOutput = exports.getDevEnvironment = exports.DevEnvironment = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.DevEnvironment = null;
utilities.lazyLoad(exports, ["DevEnvironment"], () => require("./devEnvironment"));
exports.getDevEnvironment = null;
exports.getDevEnvironmentOutput = null;
utilities.lazyLoad(exports, ["getDevEnvironment", "getDevEnvironmentOutput"], () => require("./getDevEnvironment"));
exports.Project = null;
utilities.lazyLoad(exports, ["Project"], () => require("./project"));
exports.SourceRepository = null;
utilities.lazyLoad(exports, ["SourceRepository"], () => require("./sourceRepository"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:codecatalyst/devEnvironment:DevEnvironment":
                return new exports.DevEnvironment(name, undefined, { urn });
            case "aws:codecatalyst/project:Project":
                return new exports.Project(name, undefined, { urn });
            case "aws:codecatalyst/sourceRepository:SourceRepository":
                return new exports.SourceRepository(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "codecatalyst/devEnvironment", _module);
pulumi.runtime.registerResourceModule("aws", "codecatalyst/project", _module);
pulumi.runtime.registerResourceModule("aws", "codecatalyst/sourceRepository", _module);
