// @pulumi/aws/s3control/accessGrantsInstance.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessGrantsInstance = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a resource to manage an S3 Access Grants instance, which serves as a logical grouping for access grants.
 * You can have one S3 Access Grants instance per Region in your account.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.s3control.AccessGrantsInstance("example", {});
 * ```
 *
 * ### AWS IAM Identity Center
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.s3control.AccessGrantsInstance("example", {identityCenterArn: "arn:aws:sso:::instance/ssoins-890759e9c7bfdc1d"});
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import S3 Access Grants instances using the `account_id`. For example:
 *
 * ```sh
 * $ pulumi import aws:s3control/accessGrantsInstance:AccessGrantsInstance example 123456789012
 * ```
 */
class AccessGrantsInstance extends pulumi.CustomResource {
    /**
     * Get an existing AccessGrantsInstance resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new AccessGrantsInstance(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of AccessGrantsInstance.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AccessGrantsInstance.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accessGrantsInstanceArn"] = state ? state.accessGrantsInstanceArn : undefined;
            resourceInputs["accessGrantsInstanceId"] = state ? state.accessGrantsInstanceId : undefined;
            resourceInputs["accountId"] = state ? state.accountId : undefined;
            resourceInputs["identityCenterApplicationArn"] = state ? state.identityCenterApplicationArn : undefined;
            resourceInputs["identityCenterArn"] = state ? state.identityCenterArn : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args ? args.accountId : undefined;
            resourceInputs["identityCenterArn"] = args ? args.identityCenterArn : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["accessGrantsInstanceArn"] = undefined /*out*/;
            resourceInputs["accessGrantsInstanceId"] = undefined /*out*/;
            resourceInputs["identityCenterApplicationArn"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(AccessGrantsInstance.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AccessGrantsInstance = AccessGrantsInstance;
/** @internal */
AccessGrantsInstance.__pulumiType = 'aws:s3control/accessGrantsInstance:AccessGrantsInstance';
