// @pulumi/aws/detective/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationConfiguration = exports.OrganizationAdminAccount = exports.Member = exports.InvitationAccepter = exports.Graph = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Graph = null;
utilities.lazyLoad(exports, ["Graph"], () => require("./graph"));
exports.InvitationAccepter = null;
utilities.lazyLoad(exports, ["InvitationAccepter"], () => require("./invitationAccepter"));
exports.Member = null;
utilities.lazyLoad(exports, ["Member"], () => require("./member"));
exports.OrganizationAdminAccount = null;
utilities.lazyLoad(exports, ["OrganizationAdminAccount"], () => require("./organizationAdminAccount"));
exports.OrganizationConfiguration = null;
utilities.lazyLoad(exports, ["OrganizationConfiguration"], () => require("./organizationConfiguration"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:detective/graph:Graph":
                return new exports.Graph(name, undefined, { urn });
            case "aws:detective/invitationAccepter:InvitationAccepter":
                return new exports.InvitationAccepter(name, undefined, { urn });
            case "aws:detective/member:Member":
                return new exports.Member(name, undefined, { urn });
            case "aws:detective/organizationAdminAccount:OrganizationAdminAccount":
                return new exports.OrganizationAdminAccount(name, undefined, { urn });
            case "aws:detective/organizationConfiguration:OrganizationConfiguration":
                return new exports.OrganizationConfiguration(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "detective/graph", _module);
pulumi.runtime.registerResourceModule("aws", "detective/invitationAccepter", _module);
pulumi.runtime.registerResourceModule("aws", "detective/member", _module);
pulumi.runtime.registerResourceModule("aws", "detective/organizationAdminAccount", _module);
pulumi.runtime.registerResourceModule("aws", "detective/organizationConfiguration", _module);
