// @pulumi/aws/chatbot/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsChannelConfiguration = exports.SlackChannelConfiguration = exports.getSlackWorkspaceOutput = exports.getSlackWorkspace = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.getSlackWorkspace = null;
exports.getSlackWorkspaceOutput = null;
utilities.lazyLoad(exports, ["getSlackWorkspace", "getSlackWorkspaceOutput"], () => require("./getSlackWorkspace"));
exports.SlackChannelConfiguration = null;
utilities.lazyLoad(exports, ["SlackChannelConfiguration"], () => require("./slackChannelConfiguration"));
exports.TeamsChannelConfiguration = null;
utilities.lazyLoad(exports, ["TeamsChannelConfiguration"], () => require("./teamsChannelConfiguration"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:chatbot/slackChannelConfiguration:SlackChannelConfiguration":
                return new exports.SlackChannelConfiguration(name, undefined, { urn });
            case "aws:chatbot/teamsChannelConfiguration:TeamsChannelConfiguration":
                return new exports.TeamsChannelConfiguration(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "chatbot/slackChannelConfiguration", _module);
pulumi.runtime.registerResourceModule("aws", "chatbot/teamsChannelConfiguration", _module);
