// @pulumi/aws/chatbot/slackChannelConfiguration.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlackChannelConfiguration = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS Chatbot Slack Channel Configuration.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const test = new aws.chatbot.SlackChannelConfiguration("test", {
 *     configurationName: "min-slaka-kanal",
 *     iamRoleArn: testAwsIamRole.arn,
 *     slackChannelId: "C07EZ1ABC23",
 *     slackTeamId: "T07EA123LEP",
 *     tags: {
 *         Name: "min-slaka-kanal",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import Chatbot Slack Channel Configuration using the `chat_configuration_arn`. For example:
 *
 * ```sh
 * $ pulumi import aws:chatbot/slackChannelConfiguration:SlackChannelConfiguration example arn:aws:chatbot::123456789012:chat-configuration/slack-channel/min-slaka-kanal
 * ```
 */
class SlackChannelConfiguration extends pulumi.CustomResource {
    /**
     * Get an existing SlackChannelConfiguration resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SlackChannelConfiguration(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SlackChannelConfiguration.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SlackChannelConfiguration.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["chatConfigurationArn"] = state ? state.chatConfigurationArn : undefined;
            resourceInputs["configurationName"] = state ? state.configurationName : undefined;
            resourceInputs["guardrailPolicyArns"] = state ? state.guardrailPolicyArns : undefined;
            resourceInputs["iamRoleArn"] = state ? state.iamRoleArn : undefined;
            resourceInputs["loggingLevel"] = state ? state.loggingLevel : undefined;
            resourceInputs["slackChannelId"] = state ? state.slackChannelId : undefined;
            resourceInputs["slackChannelName"] = state ? state.slackChannelName : undefined;
            resourceInputs["slackTeamId"] = state ? state.slackTeamId : undefined;
            resourceInputs["slackTeamName"] = state ? state.slackTeamName : undefined;
            resourceInputs["snsTopicArns"] = state ? state.snsTopicArns : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
            resourceInputs["userAuthorizationRequired"] = state ? state.userAuthorizationRequired : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.configurationName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'configurationName'");
            }
            if ((!args || args.iamRoleArn === undefined) && !opts.urn) {
                throw new Error("Missing required property 'iamRoleArn'");
            }
            if ((!args || args.slackChannelId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'slackChannelId'");
            }
            if ((!args || args.slackTeamId === undefined) && !opts.urn) {
                throw new Error("Missing required property 'slackTeamId'");
            }
            resourceInputs["configurationName"] = args ? args.configurationName : undefined;
            resourceInputs["guardrailPolicyArns"] = args ? args.guardrailPolicyArns : undefined;
            resourceInputs["iamRoleArn"] = args ? args.iamRoleArn : undefined;
            resourceInputs["loggingLevel"] = args ? args.loggingLevel : undefined;
            resourceInputs["slackChannelId"] = args ? args.slackChannelId : undefined;
            resourceInputs["slackTeamId"] = args ? args.slackTeamId : undefined;
            resourceInputs["snsTopicArns"] = args ? args.snsTopicArns : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["userAuthorizationRequired"] = args ? args.userAuthorizationRequired : undefined;
            resourceInputs["chatConfigurationArn"] = undefined /*out*/;
            resourceInputs["slackChannelName"] = undefined /*out*/;
            resourceInputs["slackTeamName"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(SlackChannelConfiguration.__pulumiType, name, resourceInputs, opts);
    }
}
exports.SlackChannelConfiguration = SlackChannelConfiguration;
/** @internal */
SlackChannelConfiguration.__pulumiType = 'aws:chatbot/slackChannelConfiguration:SlackChannelConfiguration';
