// @pulumi/aws/auditmanager/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationAdminAccountRegistration = exports.getFrameworkOutput = exports.getFramework = exports.getControlOutput = exports.getControl = exports.FrameworkShare = exports.Framework = exports.Control = exports.AssessmentReport = exports.AssessmentDelegation = exports.Assessment = exports.AccountRegistration = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AccountRegistration = null;
utilities.lazyLoad(exports, ["AccountRegistration"], () => require("./accountRegistration"));
exports.Assessment = null;
utilities.lazyLoad(exports, ["Assessment"], () => require("./assessment"));
exports.AssessmentDelegation = null;
utilities.lazyLoad(exports, ["AssessmentDelegation"], () => require("./assessmentDelegation"));
exports.AssessmentReport = null;
utilities.lazyLoad(exports, ["AssessmentReport"], () => require("./assessmentReport"));
exports.Control = null;
utilities.lazyLoad(exports, ["Control"], () => require("./control"));
exports.Framework = null;
utilities.lazyLoad(exports, ["Framework"], () => require("./framework"));
exports.FrameworkShare = null;
utilities.lazyLoad(exports, ["FrameworkShare"], () => require("./frameworkShare"));
exports.getControl = null;
exports.getControlOutput = null;
utilities.lazyLoad(exports, ["getControl", "getControlOutput"], () => require("./getControl"));
exports.getFramework = null;
exports.getFrameworkOutput = null;
utilities.lazyLoad(exports, ["getFramework", "getFrameworkOutput"], () => require("./getFramework"));
exports.OrganizationAdminAccountRegistration = null;
utilities.lazyLoad(exports, ["OrganizationAdminAccountRegistration"], () => require("./organizationAdminAccountRegistration"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:auditmanager/accountRegistration:AccountRegistration":
                return new exports.AccountRegistration(name, undefined, { urn });
            case "aws:auditmanager/assessment:Assessment":
                return new exports.Assessment(name, undefined, { urn });
            case "aws:auditmanager/assessmentDelegation:AssessmentDelegation":
                return new exports.AssessmentDelegation(name, undefined, { urn });
            case "aws:auditmanager/assessmentReport:AssessmentReport":
                return new exports.AssessmentReport(name, undefined, { urn });
            case "aws:auditmanager/control:Control":
                return new exports.Control(name, undefined, { urn });
            case "aws:auditmanager/framework:Framework":
                return new exports.Framework(name, undefined, { urn });
            case "aws:auditmanager/frameworkShare:FrameworkShare":
                return new exports.FrameworkShare(name, undefined, { urn });
            case "aws:auditmanager/organizationAdminAccountRegistration:OrganizationAdminAccountRegistration":
                return new exports.OrganizationAdminAccountRegistration(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "auditmanager/accountRegistration", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/assessment", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/assessmentDelegation", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/assessmentReport", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/control", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/framework", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/frameworkShare", _module);
pulumi.runtime.registerResourceModule("aws", "auditmanager/organizationAdminAccountRegistration", _module);
