// @pulumi/aws/emrcontainers/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualCluster = exports.JobTemplate = exports.getVirtualClusterOutput = exports.getVirtualCluster = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.getVirtualCluster = null;
exports.getVirtualClusterOutput = null;
utilities.lazyLoad(exports, ["getVirtualCluster", "getVirtualClusterOutput"], () => require("./getVirtualCluster"));
exports.JobTemplate = null;
utilities.lazyLoad(exports, ["JobTemplate"], () => require("./jobTemplate"));
exports.VirtualCluster = null;
utilities.lazyLoad(exports, ["VirtualCluster"], () => require("./virtualCluster"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:emrcontainers/jobTemplate:JobTemplate":
                return new exports.JobTemplate(name, undefined, { urn });
            case "aws:emrcontainers/virtualCluster:VirtualCluster":
                return new exports.VirtualCluster(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "emrcontainers/jobTemplate", _module);
pulumi.runtime.registerResourceModule("aws", "emrcontainers/virtualCluster", _module);
