// @pulumi/aws/chime/sdkvoiceSipRule.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdkvoiceSipRule = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * A SIP rule associates your SIP media application with a phone number or a Request URI hostname. You can associate a SIP rule with more than one SIP media application. Each application then runs only that rule.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.chime.SdkvoiceSipRule("example", {
 *     name: "example-sip-rule",
 *     triggerType: "RequestUriHostname",
 *     triggerValue: example_voice_connector.outboundHostName,
 *     targetApplications: [{
 *         priority: 1,
 *         sipMediaApplicationId: example_sma.id,
 *         awsRegion: "us-east-1",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import a ChimeSDKVoice SIP Rule using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:chime/sdkvoiceSipRule:SdkvoiceSipRule example abcdef123456
 * ```
 */
class SdkvoiceSipRule extends pulumi.CustomResource {
    /**
     * Get an existing SdkvoiceSipRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SdkvoiceSipRule(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of SdkvoiceSipRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SdkvoiceSipRule.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["disabled"] = state ? state.disabled : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["targetApplications"] = state ? state.targetApplications : undefined;
            resourceInputs["triggerType"] = state ? state.triggerType : undefined;
            resourceInputs["triggerValue"] = state ? state.triggerValue : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.targetApplications === undefined) && !opts.urn) {
                throw new Error("Missing required property 'targetApplications'");
            }
            if ((!args || args.triggerType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'triggerType'");
            }
            if ((!args || args.triggerValue === undefined) && !opts.urn) {
                throw new Error("Missing required property 'triggerValue'");
            }
            resourceInputs["disabled"] = args ? args.disabled : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["targetApplications"] = args ? args.targetApplications : undefined;
            resourceInputs["triggerType"] = args ? args.triggerType : undefined;
            resourceInputs["triggerValue"] = args ? args.triggerValue : undefined;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(SdkvoiceSipRule.__pulumiType, name, resourceInputs, opts);
    }
}
exports.SdkvoiceSipRule = SdkvoiceSipRule;
/** @internal */
SdkvoiceSipRule.__pulumiType = 'aws:chime/sdkvoiceSipRule:SdkvoiceSipRule';
