// @pulumi/aws/resourceexplorer/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.View = exports.searchOutput = exports.search = exports.Index = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.Index = null;
utilities.lazyLoad(exports, ["Index"], () => require("./index_"));
exports.search = null;
exports.searchOutput = null;
utilities.lazyLoad(exports, ["search", "searchOutput"], () => require("./search"));
exports.View = null;
utilities.lazyLoad(exports, ["View"], () => require("./view"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:resourceexplorer/index:Index":
                return new exports.Index(name, undefined, { urn });
            case "aws:resourceexplorer/view:View":
                return new exports.View(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "resourceexplorer/index", _module);
pulumi.runtime.registerResourceModule("aws", "resourceexplorer/view", _module);
