// @pulumi/aws/pinpoint/smsvoicev2PhoneNumber.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Smsvoicev2PhoneNumber = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Manages an AWS End User Messaging SMS phone number.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.pinpoint.Smsvoicev2PhoneNumber("example", {
 *     isoCountryCode: "US",
 *     messageType: "TRANSACTIONAL",
 *     numberType: "TOLL_FREE",
 *     numberCapabilities: ["SMS"],
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import phone numbers using the `id`. For example:
 *
 * ```sh
 * $ pulumi import aws:pinpoint/smsvoicev2PhoneNumber:Smsvoicev2PhoneNumber example phone-abcdef0123456789abcdef0123456789
 * ```
 */
class Smsvoicev2PhoneNumber extends pulumi.CustomResource {
    /**
     * Get an existing Smsvoicev2PhoneNumber resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Smsvoicev2PhoneNumber(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Smsvoicev2PhoneNumber.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Smsvoicev2PhoneNumber.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["arn"] = state ? state.arn : undefined;
            resourceInputs["deletionProtectionEnabled"] = state ? state.deletionProtectionEnabled : undefined;
            resourceInputs["isoCountryCode"] = state ? state.isoCountryCode : undefined;
            resourceInputs["messageType"] = state ? state.messageType : undefined;
            resourceInputs["monthlyLeasingPrice"] = state ? state.monthlyLeasingPrice : undefined;
            resourceInputs["numberCapabilities"] = state ? state.numberCapabilities : undefined;
            resourceInputs["numberType"] = state ? state.numberType : undefined;
            resourceInputs["optOutListName"] = state ? state.optOutListName : undefined;
            resourceInputs["phoneNumber"] = state ? state.phoneNumber : undefined;
            resourceInputs["registrationId"] = state ? state.registrationId : undefined;
            resourceInputs["selfManagedOptOutsEnabled"] = state ? state.selfManagedOptOutsEnabled : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
            resourceInputs["timeouts"] = state ? state.timeouts : undefined;
            resourceInputs["twoWayChannelArn"] = state ? state.twoWayChannelArn : undefined;
            resourceInputs["twoWayChannelEnabled"] = state ? state.twoWayChannelEnabled : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.isoCountryCode === undefined) && !opts.urn) {
                throw new Error("Missing required property 'isoCountryCode'");
            }
            if ((!args || args.messageType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'messageType'");
            }
            if ((!args || args.numberCapabilities === undefined) && !opts.urn) {
                throw new Error("Missing required property 'numberCapabilities'");
            }
            if ((!args || args.numberType === undefined) && !opts.urn) {
                throw new Error("Missing required property 'numberType'");
            }
            resourceInputs["deletionProtectionEnabled"] = args ? args.deletionProtectionEnabled : undefined;
            resourceInputs["isoCountryCode"] = args ? args.isoCountryCode : undefined;
            resourceInputs["messageType"] = args ? args.messageType : undefined;
            resourceInputs["numberCapabilities"] = args ? args.numberCapabilities : undefined;
            resourceInputs["numberType"] = args ? args.numberType : undefined;
            resourceInputs["optOutListName"] = args ? args.optOutListName : undefined;
            resourceInputs["registrationId"] = args ? args.registrationId : undefined;
            resourceInputs["selfManagedOptOutsEnabled"] = args ? args.selfManagedOptOutsEnabled : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["timeouts"] = args ? args.timeouts : undefined;
            resourceInputs["twoWayChannelArn"] = args ? args.twoWayChannelArn : undefined;
            resourceInputs["twoWayChannelEnabled"] = args ? args.twoWayChannelEnabled : undefined;
            resourceInputs["arn"] = undefined /*out*/;
            resourceInputs["monthlyLeasingPrice"] = undefined /*out*/;
            resourceInputs["phoneNumber"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Smsvoicev2PhoneNumber.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Smsvoicev2PhoneNumber = Smsvoicev2PhoneNumber;
/** @internal */
Smsvoicev2PhoneNumber.__pulumiType = 'aws:pinpoint/smsvoicev2PhoneNumber:Smsvoicev2PhoneNumber';
