// @pulumi/aws/securitylake/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriberNotification = exports.Subscriber = exports.DataLake = exports.CustomLogSource = exports.AwsLogSource = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AwsLogSource = null;
utilities.lazyLoad(exports, ["AwsLogSource"], () => require("./awsLogSource"));
exports.CustomLogSource = null;
utilities.lazyLoad(exports, ["CustomLogSource"], () => require("./customLogSource"));
exports.DataLake = null;
utilities.lazyLoad(exports, ["DataLake"], () => require("./dataLake"));
exports.Subscriber = null;
utilities.lazyLoad(exports, ["Subscriber"], () => require("./subscriber"));
exports.SubscriberNotification = null;
utilities.lazyLoad(exports, ["SubscriberNotification"], () => require("./subscriberNotification"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:securitylake/awsLogSource:AwsLogSource":
                return new exports.AwsLogSource(name, undefined, { urn });
            case "aws:securitylake/customLogSource:CustomLogSource":
                return new exports.CustomLogSource(name, undefined, { urn });
            case "aws:securitylake/dataLake:DataLake":
                return new exports.DataLake(name, undefined, { urn });
            case "aws:securitylake/subscriber:Subscriber":
                return new exports.Subscriber(name, undefined, { urn });
            case "aws:securitylake/subscriberNotification:SubscriberNotification":
                return new exports.SubscriberNotification(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "securitylake/awsLogSource", _module);
pulumi.runtime.registerResourceModule("aws", "securitylake/customLogSource", _module);
pulumi.runtime.registerResourceModule("aws", "securitylake/dataLake", _module);
pulumi.runtime.registerResourceModule("aws", "securitylake/subscriber", _module);
pulumi.runtime.registerResourceModule("aws", "securitylake/subscriberNotification", _module);
